apps/crn-server/src/controllers/research-outputs.ts

```diff
@@ -147,79 +152,82 @@ export default class ResearchOutputs implements ResearchOutputController {
+  }
 
-    const researchTags = queryResearchTagsContentsWithTotal?.items || [];
+  async update(
+    researchOutputId: string,
+    researchOutputData: ResearchOutputUpdateData,
+  ): Promise<Partial<ResearchOutputResponse>> {
+    return this.upsertResearchOutput({
+      researchOutputId,
+      ...researchOutputData,
+    });
+  }
 
-    const methods = mapResearchTags(
-      researchTags,
-      'Method',
-      researchOutputData.methods,
-      'methods',
+  private async upsertResearchOutput(
+    researchOutputData:
+      | (ResearchOutputUpdateData & { researchOutputId: string })
+      | ResearchOutputCreateData,
+  ) {
+    await this.validateResearchOutputUniques(
+      researchOutputData,
+      ('researchOutputId' in researchOutputData &&
+        researchOutputData.researchOutputId) ||
+        '',
     );
 
-    const organisms = mapResearchTags(
-      researchTags,
-      'Organism',
-      researchOutputData.organisms,
-      'organisms',
+    const researchOutputAuthors = await Promise.all(
+      (researchOutputData.authors || []).map((author) =>
+        this.mapAuthorInputToId(author),
+      ),
     );
 
-    const environments = mapResearchTags(
-      researchTags,
-      'Environment',
-      researchOutputData.environments,
-      'environments',
-    );
+    const { methods, organisms, environments, subtype } =
+      await this.parseResearchTags({ ...researchOutputData });
+
+    let researchOutputId: string;
+
+    if ('researchOutputId' in researchOutputData) {
+      const {
+        researchOutputId: _,
+        teams: __,
+        ...researchOutputUpdateData
+      } = researchOutputData;
+
+      await this.updateResearchOutput(researchOutputData.researchOutputId, {
+        ...researchOutputUpdateData,
+        authors: researchOutputAuthors,
+        methods,
+        organisms,
+        environments,
+        subtype,
+      });
 
-    const subtype = researchOutputData.subtype
-      ? [
-          mapResearchTag(
-            researchTags,
-            'Subtype',
-            researchOutputData.subtype,
-            'subtype',
-          ),
-        ]
-      : [];
+      researchOutputId = researchOutputData.researchOutputId;
+    } else {
+      const { teams, ...researchOutputCreateData } = researchOutputData;
+      const result = await this.createResearchOutput({
+        ...researchOutputCreateData,
+        authors: researchOutputAuthors,
+        methods,
+        organisms,
+        environments,
+        subtype,
+      });
 
-    const { id: researchOutputId } = await this.createResearchOutput({
-      authors: researchOutputAuthors,
-      ...researchOutputData,
-      methods,
-      organisms,
-      environments,
-      subtype,
-    });
+      researchOutputId = result.id;
 
-    await Promise.all(
-      teams.map((teamId) =>
-        this.associateResearchOutputToTeam(teamId, researchOutputId),
-      ),
-    );
+      await Promise.all(
```

Comment:
I know this isn't from this PR, but maybe one to discuss. 

We're using a Promise.all, and which will reject upon first failure.